### ISSUE TITLE
Hot restart for ASP.NET Core

### ISSUE THREAD

**Issue by mkArtakMSFT** [dotnet/aspnetcore#27270]

**Comment by Pilchie** [comment]
Is this the same as #18486?

**Comment by DamianEdwards** [comment]
No. This issue is about doing work to support restarting an ASP.NET Core application after it's been built with changes, without having to kill the application process (Hot **Restart**) whereas #18486 is about applying code changes to a running ASP.NET Core application without restarting it at all (Hot **Reload**).

Not all code changes will be supported by Hot Reload, and thus to improve the inner-loop experience for developers when those changes occur we can make ASP.NET Core able to restart much faster than today.

**Comment by hiredgunhouse** [comment]
You'll need to start with basics like shadow copying of assemblies cause a started ASP.NET Core app seems to lock files on disk, which is not only a problem for this feature but also for an ability to work in a sane manner within a monorepo.
I think while .NET Core was a big and much needed step forward you have still managed made a lot of mistakes that could be avoided if you only spent a bit of time thinking about consequences, but hey you're not slowing down and with .NET 5 being still relatively fresh we're talking .NET 6 and 7, 8 and 9 coming probably pretty soon. While it's good you improve things this also creates a lot of friction for us developers on the ground trying to do real projects and work for our paying clients. My last two projects started on the latest .NET version 2.1 and by the time it was done we had to migrate to .NET 3, the same happened for another project that started on .NET 3 and soon .NET 5 came out and we had to migrate just to make sure we're not going with "legacy" stuff into production. I wish you just had spent more time doing some upfront design and validate that design and gather feedback and have longer alpha and beta cycles, I'm sure that would save us from the WebHost vs Generic Host pains and you would realize quicker that it makes sense to do things like Configure overloads for environments, etc. Oh, and the JSON serializer? Why couldn't you make it more on par with Newtonsoft instead of releasing it early?
And now the hot reloading, I have to admit I wasn't expecting MS to actually realize this is so important so "relatively" quickly. So I guess kudos to smart people like you who are probably the champions of those ideas there. You will face a big challenge there because of the historical lack of oversight and vision in the architecture and design of .NET and C# as a language, the general direction of the platform and language where there has historically been little space for things like immutability and pure functions will be the extra tax you will have pay and will make your life really hard because just reloading the code without ensuring state is not lost will not give you any decent results and the much hated JavaScript has a lot of advantages there and is currently the cutting edge and and leading place of the progress in the area that you will have a hard time to even replicate cause React embraces FP and externalizes state thanks hooks which both contribute to great hot reloading experience so I guess you should look for inspiration there.
Anyway, a long and complicated subject, glad you are finally starting that important work and good luck!
